addons/stock/models/stock_quant.py

```diff
@@ -127,6 +132,50 @@ def _compute_inventory_date(self):
         for quant in quants:
             quant.inventory_date = date_by_location[quant.location_id]
 
+    def _compute_last_count_date(self):
+        """ We look at the stock move lines associated with every quant to get the last count date.
+        """
+        groupby_vars = ['sml.product_id', 'sml.lot_id', 'sml.company_id', 'sml.location_id', 'sml.location_dest_id', 'sml.package_id', 'sml.result_package_id']
+        where_clause = f"""
+            move.is_inventory is true AND
+            sml.state = 'done' AND
+            sml.product_id in ({', '.join([str(x) for x in self.product_id.ids])}) AND
+            {'sml.lot_id in (' + ', '.join([str(x) for x in self.lot_id.ids]) + ') OR ' if self.lot_id else ''} lot_id is null AND
+            sml.company_id in ({', '.join([str(x) for x in self.company_id.ids])}) AND
+            sml.location_id in ({', '.join([str(x) for x in self.location_id.ids])}) OR
+            sml.location_dest_id in ({', '.join([str(x) for x in self.location_id.ids])}) AND
+            {'sml.package_id in (' + ', '.join([str(x) for x in self.package_id.ids]) + ') OR ' if self.package_id else ''} package_id is null OR
+            {'sml.result_package_id in (' + ', '.join([str(x) for x in self.package_id.ids]) + ') OR ' if self.package_id else ''} result_package_id is null
+        """
+        query = f"""
+               SELECT max(sml.date), {', '.join(groupby_vars)}
+                 FROM stock_move_line sml
+                 JOIN stock_move move ON sml.move_id = move.id
+                WHERE {where_clause}
+             GROUP BY {', '.join(groupby_vars)}
+         """
+        self.env.cr.execute(query)
+        results = self.env.cr.fetchall()
+        results_dic = [{var: result[idx] for idx, var in enumerate(['date'] + groupby_vars)} for result in results]
+
+        # We need to match the quant location with either the location_id or the dest_location_id of the move line (same for packages)
+        results_merged = {}
+        for key_to_remove1 in ['sml.location_id', 'sml.dest_location_id']:
+            for key_to_remove2 in ['sml.package_id', 'sml.result_package_id']:
+                res_without_vars = [{k: v for k, v in d.items() if k not in (key_to_remove1, key_to_remove2)} for d in results_dic]
+                for res in res_without_vars:
+                    date = res.pop('date')
+                    results_merged.update({tuple(res.values()): date})
+
+        for quant in self:
+            quant.last_count_date = results_merged.get((
+                quant.product_id.id,
+                quant.lot_id.id or None,
+                quant.company_id.id,
+                quant.location_id.id,
+                quant.package_id.id or None,
+            ))
+
```

Comment:
```suggestion
        self.last_count_date = False
        groups = self.env['stock.move.line'].read_group(
            [
                ('state', '=', 'done'),
                ('move_id.is_inventory', '=', True),
                '|',
                    ('lot_id', 'in', self.lot_id.ids),
                    ('lot_id', '=', False),
                '|',
                    ('owner_id', 'in', self.owner_id.ids),
                    ('owner_id', '=', False),
                '|',
                    ('location_id', 'in', self.location_id.ids),
                    ('location_dest_id', 'in', self.location_id.ids),
                '|',
                    ('package_id', '=', False),
                    '|',
                        ('package_id', 'in', self.package_id.ids),
                        ('result_package_id', 'in', self.package_id.ids),

            ],
            ['date:max', 'lot_id', 'package_id', 'owner_id', 'result_package_id', 'location_id', 'location_dest_id'],
            ['lot_id', 'package_id', 'owner_id', 'result_package_id', 'location_id', 'location_dest_id'],
            lazy=False)

        def _update_dict(date_by_quant, key, value):
            current_date = date_by_quant.get(False)
            if not current_date or value > current_date:
                date_by_quant[key] = value

        date_by_quant = {}
        for group in groups:
            move_line_date = group['date']
            location_id, location_dest_id, package_id, result_package_id, lot_id, owner_id = (
                group['location_id'][0], group['location_dest_id'][0], group['package_id'] and group['package_id'][0],
                group['result_package_id'] and group['result_package_id'][0], group['lot_id'] and group['lot_id'][0],
                group['owner_id'] and group['owner_id'][0])
            _update_dict(date_by_quant, (location_id, package_id, lot_id, owner_id), move_line_date)
            _update_dict(date_by_quant, (location_dest_id, package_id, lot_id, owner_id), move_line_date)
            _update_dict(date_by_quant, (location_id, result_package_id, lot_id, owner_id), move_line_date)
            _update_dict(date_by_quant, (location_dest_id, result_package_id, lot_id, owner_id), move_line_date)

        for quant in self:
            quant.last_count_date = date_by_quant.get((quant.location_id.id, quant.package_id.id, quant.lot_id.id, quant.owner_id.id))
```
What about that solution. It would avoid a SQL query and it would result the same